pkg/config/manifest.go

```diff
@@ -566,220 +568,73 @@ func (s *Step) GetMixinName() string {
-	}
-
-	sourceType := source[1]
-	sourceName := source[2]
-
-	var replacement interface{}
-	switch sourceType {
-	case "parameters":
-		for _, param := range m.Parameters {
-			if param.Name == sourceName {
-				if param.Destination == nil {
-					// Porter by default sets CNAB params to name.ToUpper()
-					pe := strings.ToUpper(sourceName)
-					replacement = os.Getenv(pe)
-				} else if param.Destination.EnvironmentVariable != "" {
-					replacement = os.Getenv(param.Destination.EnvironmentVariable)
-				} else if param.Destination == nil && param.Destination.Path != "" {
-					replacement = param.Destination.Path
-				} else {
-					return nil, errors.New(
-						"unknown parameter definition, no environment variable or path specified",
-					)
-				}
-				// if replacement has been set and parameter is designated sensitive, add to list of sensitive values
-				if replacement != nil && param.Sensitive {
-					m.sensitiveValues = append(m.sensitiveValues, reflect.ValueOf(replacement).String())
-				}
-			}
-		}
-	case "credentials":
-		for _, cred := range m.Credentials {
-			if cred.Name == sourceName {
-				if cred.Path != "" {
-					replacement = cred.Path
-				} else if cred.EnvironmentVariable != "" {
-					replacement = os.Getenv(cred.EnvironmentVariable)
-				} else {
-					return nil, errors.New(
-						"unknown credential definition, no environment variable or path specified",
-					)
-				}
+// ResolveStep will walk through the Step's data and resolve any placeholder
+// data using the definitions in the manifest, like parameters or credentials.
+func (m *Manifest) ResolveStep(step *Step) error {
 
-				// if replacement has been set, add to list of sensitive values
-				if replacement != nil {
-					m.sensitiveValues = append(m.sensitiveValues, reflect.ValueOf(replacement).String())
-				}
-			}
-		}
-	case "outputs":
-		if o, exists := m.outputs[sourceName]; exists {
-			replacement = o
-			// Porter considers all outputs as sensitive
-			if replacement != nil {
-				m.sensitiveValues = append(m.sensitiveValues, reflect.ValueOf(replacement).String())
-			}
-		}
-	case "dependencies":
-		for _, dep := range m.Dependencies {
-			if dep.Name == sourceName {
-				replacement, err := dep.resolveValue(key)
-				// Retrieve updated list of sensitive values from dependency and add to our list
-				for _, val := range dep.m.GetSensitiveValues() {
-					m.sensitiveValues = append(m.sensitiveValues, val)
-				}
-				return replacement, err
-			}
-		}
-	default:
-		return nil, errors.New(fmt.Sprintf("unknown source specification: %s", key))
+	mustache.AllowMissingVariables = false
+	sourceData, err := m.buildSourceData()
+	if err != nil {
+		return errors.Wrap(err, "unable to resolve step")
 	}
-
-	if replacement == nil {
-		return nil, errors.New(fmt.Sprintf("no value found for source specification: %s", key))
+	payload, err := yaml.Marshal(step)
+	if err != nil {
+		return errors.Wrap(err, "unable to resolve step")
```

Comment:
I'm not sure wrapping all these various errors with the same error message is going to be helpful? Having a bit more context about what was going on may be useful when dealing with errors like "unable to resolve step: unexpected EOF", you know?